### ISSUE TITLE
The meaning of  norm_val

### ISSUE THREAD

**Issue by tylerswift6** [XMed-Lab/AdaCon#3]
Super valuable work! 
I have a question. What's the meaning of  the parameter norm_val and how to set it? is 2/ batch_size ?
and what's the difference between outcome and outcome_cls?

Thank you very much!

**Comment by ackbar03** [comment]
Hi, and thanks for your interest!

`what's the difference between outcome and outcome_cls?`

Outcome is the LVEF label. 

outcome_cls is calculated from the dataset class, lines [108 - 123 ](https://github.com/xmed-lab/AdaCon/blob/d52785ac7001fbbab4b165c675e674fb4af20cfc/echonet/datasets/echo.py#L108). It is essentially the cumulative number of samples with LVEF lower than the LVEF label

`What's the meaning of the parameter norm_val and how to set it? is 2/ batch_size ?`

norm_val is equal to 2/7465 which is 2/(total number of samples in training set). This essentially normalizes the value of outcome_cls to range (0,2) in the loss function, [line 543](https://github.com/xmed-lab/AdaCon/blob/d52785ac7001fbbab4b165c675e674fb4af20cfc/echonet/models/rnet2dp1.py#L543) such that the margin is 2*[ecdf(y_i) - ecdf(y_j)] as defined in the paper. 

The code was originally written this way to facilitate the ablation experiments I was running at the time. I may re-write it at a later time to make it easier to comprehend. 

Hope that helps answer your question!

**Comment by tylerswift6** [comment]
Thank you for your quick reply!

For the outcome_cls, you write
loss_ctr = criterion_cntrst(features, outcome_cls, dist = outcome_cls, norm_val=2/7465, scale_s = 150)

so the dist is equal to label,
Could I understand the outcome_cls is the ground truth lables for one batch of images(because I am a little confused the LVEF lable, I am familiar with the object detection, so sorry)

**Comment by ackbar03** [comment]
Sorry for the much later reply

LVEF label is basically the label given by the dataset and the one we are trying to predict, `y`

> Could I understand the outcome_cls is the ground truth lables for one batch of images

This is incorrect.

`outcome_cls` for sample `i` is 
`sum(number of samples which has label < y_i`
So it is basically the cumulative number of samples, within the entire training set with LVEF label less than `y_i` (LVEF or EF is the label that comes with the EchoNet-Dynamic dataset and is what we are trying to predict)

So it is basically ECDF(x_i) * Total_number_of_samples
Where ECDF is the empirical distribution function. 

This is then fed into the `criterion_cntrst` where it is multiplied by the normalizing factor 2/7465 at one point to get the AdaCon loss function for calculation. 

You'll have to go through the code to see how the values get manipulated but hopefully it shouldn't be too difficult to match with the operations described in the paper. You can add some print statements to help debug the flow. 

Hope this helps somewhat.
